src/common/scratchpad.cpp

```diff
@@ -74,14 +74,14 @@ struct global_scratchpad_t : public scratchpad_t {
     virtual char *get() const { return scratchpad_; }
 
 private:
-    thread_local static char *scratchpad_;
-    thread_local static size_t size_;
-    thread_local static unsigned int reference_count_;
+    static char *scratchpad_;
+    static size_t size_;
+    static unsigned int reference_count_;
 };
 
-thread_local char *global_scratchpad_t::scratchpad_ = nullptr;
-thread_local size_t global_scratchpad_t::size_ = 0;
-thread_local unsigned int global_scratchpad_t::reference_count_ = 0;
+char *global_scratchpad_t::scratchpad_ = nullptr;
+size_t global_scratchpad_t::size_ = 0;
+unsigned int global_scratchpad_t::reference_count_ = 0;
```

Comment:
This behavior of the `global_scratchpad` in intended.
The global scratchpad is not thread safe, so creating a primitive using the global scratchpad in one thread and executing it in another fails (whether the buffer is `thread_local` or not).
The `thread_local` buffers enable some level of concurrency though (it works when you spawn multiple threads, and each thread creates and runs its own primitives). 
When creating and executing primitives in different threads, the user can either use concurrent scratchpad or pass the scratchpad as a parameter to the primitive execution.

In the example you mention (1 thread creates primitives, 2 threads execute them), it is true that you avoid a segfault, but the result will be incorrect (since two primitives are concurrently accessing the same buffer). For such a scenario, the global scratchpad should not be used.
However, if we remove the `thread_local`, it can cause segfaults in a currently supported scenario (2 threads, each creating and executing their own primitives).

Reply:
I think you misunderstood the issue. You don't need concurrent execution at all to cause a crash with the current behavior. There's no concurrency, no multiple threads executing at the same time. The crash happens if you initialize a primitive on thread A, and later invoke that primitive from thread B, but thread A and thread B are *not* concurrent, there's no issue of thread safety, everything is synchronized. Removing thread_local does *not* cause a crash in this scenario, and this is how Open Image Denoise is currently working. I think this is quite dangerous behavior and took quite some time to figure out why a segfault is happening. If multiple threads want to execute primitives, wouldn't a concurrent scratchpad be the right option instead?

Reply:
Out of curiosity, why do you create the primitives and execute them in separate threads if there is no concurrency?

Reply:
It's not Open Image Denoise what directly invokes this from different threads, but other libraries and applications that use Open Image Denoise. E.g. OSPRay (http://www.ospray.org/), which executes denoising asynchronously with other rendering operations on a thread different from the main thread. So actually there is concurrency, but there is always only one thread invoking MKL-DNN operations at a time. The other threads are executing other tasks.

Reply:
Currently, if scratchpad mode is "library" then a user has to create and execute a primitive in the same thread otherwise it can crash if the primitive uses global scratchpad. 
This change also makes primitive creation not thread-safe.

Reply:
Yes, but it shouldn't crash if there's no concurrency.

Reply:
I would suggest for OID to switch to user scratchpads. This was not possible in 0.x, but no there's no reason not to if the usage model is different from what the library supports.

Reply:
Agree, it shouldn't crash with this change, but it fixes one thing and breaks another. I agree with @rsdubtso on his suggestion to use user scratchpad mode.

Reply:
We've already switched to user scratchpads, but I still think the current behavior of MKL-DNN is very confusing and error-prone. Is this documented somewhere?

Reply:
Scratchpad has it's own section in the [developer guide](http://intel.github.io/mkl-dnn/dev_guide_attributes_scratchpad.html), however the scenario you have is not covered there. I agree that this behavior should be documented.

Reply:
Scratchpad section in [developer guide](http://intel.github.io/mkl-dnn/dev_guide_attributes_scratchpad.html) is extended with additional information on thread safety.